Fix bug where canvas culls things at origin with size 0

nathanfranke opened the pull request:
Fixes #17917 

Rather than detecting if the control rectangle is intersecting, detect if it is intersecting or touching.  
It should be noted that I added another function to `Rect2` which may not be favourable

Note: Special thanks to @clayjohn for helping me find where the culling is, otherwise I couldn't have made this change

*In my opinion, this should be cherry-picked for the first maintenance release of Godot 3.2*

clayjohn commented:
I wonder if the ``intersects_touch`` would be useful in other places where ``intersects`` is called? We may be able to just update the function instead of making a new one.

nathanwfranke commented:
> I wonder if the `intersects_touch` would be useful in other places where `intersects` is called?

That is a good question. It should be noted though that the name `intersects` implies that touching would return false. We could name a method `collides`, but I don't like that since it doesn't explicitly state what would happen to adjacent rectangles, and it also implies that Rect2 is some sort of physics body

> We may be able to just update the function instead of making a new one.

Initially, I considered adding a parameter `p_touch_intersection = false` to keep code simple and not break compatibility. However, I chose not to because while minor, an extra if statement can slow down code on such a low level function.

clayjohn commented:
I think this should be fine. Before it is merged, can you squash your commits into 1 (http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase)

Commit:
Fix bug where Control at origin with 0 size not rendered


Make a new method instead to make the code more elegant


Move Function down a bit

nathanwfranke commented:
Just a final thought, should the function be named `intersects_touch` or `touches`

clayjohn commented:
In general ``touches`` is better. However, in this case, it is just a slight modification to ``intersects`` so ``intersects_touches`` is more clear about what it does. i.e. it returns true if it intersects _or_ touches.

nathanwfranke commented:
In that case, this should be ready to merge

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.2.1.
